File: templates/problems/10_backtracking/07_palindrome_partition/palindrome_partition_test.go
package palindrome_partition

import (
	"sort"
	"testing"
)

func sortStringSlices(result [][]string) {
	for _, s := range result {
		sort.Strings(s)
	}
	sort.Slice(result, func(i, j int) bool {
		a, b := result[i], result[j]
		if len(a) != len(b) {
			return len(a) < len(b)
		}
		for k := 0; k < len(a); k++ {
			if a[k] != b[k] {
				return a[k] < b[k]
			}
		}
		return false
	})
}

func equalStringSlices(a, b [][]string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if len(a[i]) != len(b[i]) {
			return false
		}
		for j := range a[i] {
			if a[i][j] != b[i][j] {
				return false
			}
		}
	}
	return true
}

func TestPartition(t *testing.T) {
	tests := []struct {
		name string
		s    string
		want [][]string
	}{
		{
			name: "example aab",
			s:    "aab",
			want: [][]string{{"a", "a", "b"}, {"aa", "b"}},
		},
		{
			name: "single char",
			s:    "a",
			want: [][]string{{"a"}},
		},
		{
			name: "all same chars",
			s:    "aaa",
			want: [][]string{{"a", "a", "a"}, {"a", "aa"}, {"aa", "a"}, {"aaa"}},
		},
		{
			name: "no palindrome partitions beyond singles",
			s:    "abc",
			want: [][]string{{"a", "b", "c"}},
		},
		{
			name: "full palindrome",
			s:    "aba",
			want: [][]string{{"a", "b", "a"}, {"aba"}},
		},
		{
			name: "two chars same",
			s:    "bb",
			want: [][]string{{"b", "b"}, {"bb"}},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := partition(tt.s)
			sortStringSlices(got)
			sortStringSlices(tt.want)
			if !equalStringSlices(got, tt.want) {
				t.Errorf("partition(%q) = %v, want %v", tt.s, got, tt.want)
			}
		})
	}
}

File: templates/problems/10_backtracking/07_palindrome_partition/typescript/palindromePartition.test.ts
import { describe, it, expect } from 'vitest';
import { partition } from './palindromePartition';

function sortStringSlices(result: string[][]): string[][] {
    for (const s of result) {
        s.sort();
    }
    result.sort((a, b) => {
        if (a.length !== b.length) return a.length - b.length;
        for (let k = 0; k < a.length; k++) {
            if (a[k] !== b[k]) return a[k] < b[k] ? -1 : 1;
        }
        return 0;
    });
    return result;
}

describe('PalindromePartition', () => {
    it('example aab', () => {
        const got = partition('aab');
        const expected = [['a', 'a', 'b'], ['aa', 'b']];
        expect(sortStringSlices(got)).toEqual(sortStringSlices(expected));
    });

    it('single char', () => {
        const got = partition('a');
        const expected = [['a']];
        expect(sortStringSlices(got)).toEqual(sortStringSlices(expected));
    });

    it('all same chars', () => {
        const got = partition('aaa');
        const expected = [['a', 'a', 'a'], ['a', 'aa'], ['aa', 'a'], ['aaa']];
        expect(sortStringSlices(got)).toEqual(sortStringSlices(expected));
    });

    it('no palindrome partitions beyond singles', () => {
        const got = partition('abc');
        const expected = [['a', 'b', 'c']];
        expect(sortStringSlices(got)).toEqual(sortStringSlices(expected));
    });

    it('full palindrome', () => {
        const got = partition('aba');
        const expected = [['a', 'b', 'a'], ['aba']];
        expect(sortStringSlices(got)).toEqual(sortStringSlices(expected));
    });

    it('two chars same', () => {
        const got = partition('bb');
        const expected = [['b', 'b'], ['bb']];
        expect(sortStringSlices(got)).toEqual(sortStringSlices(expected));
    });
});
